core/kernel/shared/sprintf_encoding.rb

```diff
@@ -0,0 +1,32 @@
+describe :kernel_sprintf_encoding, shared: true do
+  def format(*args)
+    @method.call(*args)
+  end
+
+  it "returns a String in the same encoding as the format String if compatible" do
+    string = "%s".force_encoding(Encoding::KOI8_U)
+    result = format(string, "dogs")
+    result.encoding.should equal(Encoding::KOI8_U)
+  end
+
+  it "returns a String in the argument's encoding if format encoding is more restrictive" do
+    string = "foo %s".force_encoding(Encoding::US_ASCII)
+    argument = "b\303\274r".force_encoding(Encoding::UTF_8)
+
+    result = format(string, argument)
+    result.encoding.should equal(Encoding::UTF_8)
+  end
+
+  # strange behavior
+  # irb(main):020:0* sprintf "Ä %s".encode('windows-1252'), "Ђ".encode('utf-8')
+  #   Encoding::CompatibilityError: incompatible character encodings: Windows-1252 and UTF-8
+  # expected to return string in UTF-8 encoding
+  it "raises Encoding::CompatibilityError if cannot convert argument and format to any of encodings" do
```

Comment:
@eregon There is another example of strange behavior

Reply:
Mmh, so those encoding are compatible according to Encoding.compatible?:
```ruby
> Encoding.compatible? "windows-1252", "UTF-8"
=> #<Encoding:UTF-8>
```

But `"Ђ"` cannot be encoded in Windows-1252:
```ruby
[13] pry(main)> "Ђ".encode('windows-1252')
Encoding::UndefinedConversionError: U+0402 to WINDOWS-1252 in conversion from UTF-8 to WINDOWS-1252
from (pry):12:in `encode'
```

Reply:
@eregon right,

Feature is
`it raises Encoding::CompatibilityError if cannot convert argument and format to any of encodings`.

But in my example in comments Ruby can convert format string (in 'windows-1252') to utf-8 but doesn't do this.

Reply:
I think it's basically a consequence that
```ruby
> "Ä %s".encode('windows-1252') << "Ђ".encode('utf-8')
Encoding::CompatibilityError: incompatible character encodings: Windows-1252 and UTF-8
```
I think Ruby never tries to re-encode a String, unless you use `encode` explicitly.
The resulting encoding seems therefore to be the resulting encoding of << each `%s` argument .

Reply:
@eregon maybe, but it looks like a bug to me, because [here](https://github.com/ruby/spec/pull/537/files/b00ae735141b6ef6228376a3ebdb0240c0fabdc0#diff-cf2ddc1bab1f1f37bd1ec13bcab65b60R12) this feature works

Reply:
But here the `"foo %s"` is ASCII-only, so it works fine.
There is special treatment for Strings with only 7-bit characters for many String operations (in that case the Encoding usually doesn't matter, it's like if it was US-ASCII).

Reply:
@eregon right